File: apps/lazuar-api/src/rails/razorpay_webhook.rs
//! Port of `Rails/Razorpay/RazorpayWebhook.cs`.
//! Issue 018: the rotating `X-Razorpay-Event-Id` header is the dedupe id when
//! present — replaying a captured event with a fresh header id must NOT book a
//! second fulfillment, and the header id (not the payment id) is what the
//! `(Org, Provider, EventId)` dedupe sees.

use hmac::{Hmac, Mac};
use serde_json::Value;
use sha2::Sha256;
use subtle::ConstantTimeEq;

use crate::domain::currency;
use crate::secrets::SecretBox;
use crate::webhooks::psp_parse::{Headers, ParsedWebhook, WebhookParseError};

pub const SIGNATURE_HEADER: &str = "X-Razorpay-Signature";
pub const EVENT_ID_HEADER: &str = "X-Razorpay-Event-Id";

fn header<'a>(headers: &'a Headers, name: &str) -> Option<&'a str> {
    headers.get(name).map(str::trim).filter(|v| !v.is_empty())
}

pub fn parse(
    raw: &str,
    headers: &Headers,
    cred_webhook_ciphertext: Option<&str>,
    box_one: &SecretBox,
) -> Result<ParsedWebhook, WebhookParseError> {
    let Some(ct) = cred_webhook_ciphertext.map(str::trim).filter(|s| !s.is_empty()) else {
        return Err(WebhookParseError::MissingSecret("webhook secret missing".into()));
    };

    let Some(signature) = header(headers, SIGNATURE_HEADER).filter(|s| !s.is_empty()) else {
        return Err(WebhookParseError::verify_error("invalid signature"));
    };

    let secret = box_one.unprotect(ct).map_err(|_| WebhookParseError::verify_error("invalid signature"))?;
    let mut mac = <Hmac<Sha256> as Mac>::new_from_slice(secret.as_bytes())
        .expect("hmac accepts any key length");
    mac.update(raw.as_bytes());
    let expected = hex::encode(mac.finalize().into_bytes());
    let provided = signature.trim().to_lowercase();
    if !bool::from(provided.as_bytes().ct_eq(expected.as_bytes())) {
        return Err(WebhookParseError::verify_error("invalid signature"));
    }

    let root: Value = serde_json::from_str(raw)
        .map_err(|_| WebhookParseError::verify_error("invalid event"))?;
    let event_type = root.get("event").and_then(Value::as_str);
    let entity = root.pointer("/payload/payment/entity");
    let payment_id = entity.and_then(|e| e.get("id")).and_then(Value::as_str).map(str::to_string);
    let header_event_id = header(headers, EVENT_ID_HEADER).map(str::to_string);

    if event_type == Some("payment.failed") {
        let failed_id = header_event_id.clone().or_else(|| {
            payment_id.as_deref().filter(|p| !p.trim().is_empty()).map(|p| format!("failed:{p}"))
        });
        let Some(failed_id) = failed_id else {
            return Err(WebhookParseError::verify_error("missing event id"));
        };
        let failed_checkout = entity
            .and_then(|e| e.get("notes"))
            .and_then(|n| n.get("checkout_id"))
            .and_then(Value::as_str)
            .map(str::to_string);
        return Ok(ParsedWebhook {
            event_id: failed_id,
            failed: true,
            ignore_reason: Some("payment_failed".into()),
            checkout_id: failed_checkout,
            provider_ref: payment_id,
            ..ParsedWebhook::default()
        });
    }

    let is_captured = event_type == Some("payment.captured");
    if matches!(
        event_type,
        Some("payment_link.paid") | Some("payment_link.expired") | Some("order.paid")
    ) || !is_captured
    {
        let other_id = header_event_id.clone().unwrap_or_else(|| {
            match payment_id.as_deref().filter(|p| !p.trim().is_empty()) {
                Some(p) => format!("{}:{p}", event_type.unwrap_or("razorpay")),
                None => format!("{}:none", event_type.unwrap_or("razorpay")),
            }
        });
        return Ok(ParsedWebhook {
            event_id: other_id,
            ignored: true,
            ignore_reason: event_type.map(str::to_string),
            ..ParsedWebhook::default()
        });
    }

    let Some(entity) = entity else {
        return Err(WebhookParseError::verify_error("missing payment id"));
    };
    let Some(payment_id) = payment_id else {
        return Err(WebhookParseError::verify_error("missing payment id"));
    };

    let currency = entity
        .get("currency")
        .and_then(Value::as_str)
        .and_then(|c| currency::try_normalize_currency(Some(c)))
        .ok_or_else(|| WebhookParseError::verify_error("missing currency"))?;

    // Payment entity amount is already minor (paise/sen). RM10.00 → 1000.
    let amount = entity.get("amount").and_then(Value::as_i64).unwrap_or(0);

    let checkout_id = entity
        .get("notes")
        .and_then(|n| n.get("checkout_id"))
        .and_then(Value::as_str)
        .map(str::to_string);

    let hosted_session_id = root
        .pointer("/payload/payment_link/entity/id")
        .and_then(Value::as_str)
        .map(str::to_string);

    let event_id = match header_event_id {
        Some(id) if !id.trim().is_empty() => id,
        _ => format!("captured:{payment_id}"),
    };

    Ok(ParsedWebhook {
        event_id,
        checkout_id,
        hosted_session_id,
        provider_ref: Some(payment_id),
        amount_minor: Some(amount),
        currency: Some(currency),
        ..ParsedWebhook::default()
    })
}

File: apps/lazuar-pay/src/Lazuar.Pay/Rails/Razorpay/RazorpayWebhook.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Lazuar.Pay.Data;
using Lazuar.Pay.Money;
using Lazuar.Pay.Secrets;

using Lazuar.Pay.Rails;

using Lazuar.Pay.Webhooks;

namespace Lazuar.Pay.Rails.Razorpay;

internal static class RazorpayWebhook
{
    public static PspParseResult Parse(string raw, IHeaderDictionary headers, GatewayCredentialRow cred, SecretBox box)
    {
        if (string.IsNullOrWhiteSpace(cred.WebhookCiphertext))
        {
            throw new InvalidOperationException("webhook secret missing");
        }

        var sigKey = headers.Keys.FirstOrDefault(k => k.Equals("X-Razorpay-Signature", StringComparison.OrdinalIgnoreCase));
        if (sigKey is null || !headers.TryGetValue(sigKey, out var signature) || string.IsNullOrWhiteSpace(signature))
        {
            throw new PspVerifyException("invalid signature");
        }

        var secret = box.Unprotect(cred.WebhookCiphertext);
        var mac = HMACSHA256.HashData(Encoding.UTF8.GetBytes(secret), Encoding.UTF8.GetBytes(raw));
        var expected = Convert.ToHexString(mac).ToLowerInvariant();
        var provided = signature.ToString().Trim().ToLowerInvariant();
        var left = Encoding.UTF8.GetBytes(provided);
        var right = Encoding.UTF8.GetBytes(expected);
        if (left.Length != right.Length || !CryptographicOperations.FixedTimeEquals(left, right))
        {
            throw new PspVerifyException("invalid signature");
        }

        using var doc = JsonDocument.Parse(raw);
        var eventType = doc.RootElement.TryGetProperty("event", out var ev) ? ev.GetString() : null;
        JsonElement entity = default;
        var hasEntity = doc.RootElement.TryGetProperty("payload", out var payload)
                        && payload.TryGetProperty("payment", out var payment)
                        && payment.TryGetProperty("entity", out entity);
        var paymentId = hasEntity && entity.TryGetProperty("id", out var pid) ? pid.GetString() : null;
        var headerEventId = Header(headers, "X-Razorpay-Event-Id");

        if (eventType == "payment.failed")
        {
            var failedId = headerEventId ?? (string.IsNullOrWhiteSpace(paymentId) ? null : "failed:" + paymentId);
            if (string.IsNullOrWhiteSpace(failedId))
            {
                throw new PspVerifyException("missing event id");
            }

            string? failedCheckout = null;
            if (hasEntity && entity.TryGetProperty("notes", out var failNotes) && failNotes.ValueKind == JsonValueKind.Object
                && failNotes.TryGetProperty("checkout_id", out var failCid))
            {
                failedCheckout = failCid.GetString();
            }

            return new PspParseResult
            {
                EventId = failedId,
                Failed = true,
                IgnoreReason = "payment_failed",
                CheckoutId = failedCheckout,
                ProviderRef = paymentId
            };
        }

        // Issue 004 (issues/003): Pay mints payment links, and Razorpay's own payment-links
        // docs point merchants at payment_link.paid — the event Pay used to file under
        // "ignored", so a merchant subscribing to exactly that event never saw a
        // fulfillment. The payment entity below carries everything the captured path reads
        // (id, minor amount, currency, notes, and payment_link.entity.id for session
        // binding). payment_link.expired stays ignored: the local TTL sweep owns expiry,
        // and a capture can still trail an expired link — that arrival must find the
        // late-pay route, which ignoring keeps reachable.
        if (eventType is not "payment.captured" and not "payment_link.paid")
        {
            var otherId = headerEventId
                          ?? (string.IsNullOrWhiteSpace(paymentId) ? (eventType ?? "razorpay") + ":none" : eventType + ":" + paymentId);
            return new PspParseResult { EventId = otherId, Ignored = true, IgnoreReason = eventType };
        }

        if (!hasEntity || string.IsNullOrWhiteSpace(paymentId))
        {
            throw new PspVerifyException("missing payment id");
        }

        if (!MoneyMath.TryNormalizeCurrency(
                entity.TryGetProperty("currency", out var cur) && cur.ValueKind == JsonValueKind.String ? cur.GetString() : null,
                out var currency))
        {
            throw new PspVerifyException("missing currency");
        }

        // Payment entity amount is already minor (paise/sen). RM10.00 → 1000.
        var amount = entity.TryGetProperty("amount", out var amt) && amt.ValueKind == JsonValueKind.Number
            ? amt.GetInt64()
            : 0L;

        string? checkoutId = null;
        if (entity.TryGetProperty("notes", out var notes) && notes.ValueKind == JsonValueKind.Object
            && notes.TryGetProperty("checkout_id", out var cid))
        {
            checkoutId = cid.GetString();
        }

        string? hostedSessionId = null;
        if (doc.RootElement.TryGetProperty("payload", out var pl)
            && pl.TryGetProperty("payment_link", out var link)
            && link.TryGetProperty("entity", out var linkEntity)
            && linkEntity.TryGetProperty("id", out var linkId))
        {
            hostedSessionId = linkId.GetString();
        }

        // Distinct namespace per success event type: a merchant enabling both events gets
        // two deliveries for one payment. Without the header event id they must not dedupe
        // against each other — the first fulfills and Handle answers the second benignly.
        var eventId = !string.IsNullOrWhiteSpace(headerEventId)
            ? headerEventId
            : (eventType == "payment_link.paid" ? "link_paid:" : "captured:") + paymentId;
        return new PspParseResult
        {
            EventId = eventId,
            CheckoutId = checkoutId,
            HostedSessionId = hostedSessionId,
            ProviderRef = paymentId,
            AmountMinor = amount,
            Currency = currency
        };
    }

    static string? Header(IHeaderDictionary headers, string name)
    {
        foreach (var key in headers.Keys)
        {
            if (key.Equals(name, StringComparison.OrdinalIgnoreCase))
            {
                var v = headers[key].ToString();
                return string.IsNullOrWhiteSpace(v) ? null : v.Trim();
            }
        }

        return null;
    }
}
